setup_mysql.sh

```diff
@@ -0,0 +1,230 @@
+  expect eof
+  exit [lindex [wait] 3]
+HERE
+  then
+    echo
+    echo "ROOT password assigned in mysql_pw.txt WORKED! MySQL is installed & mysql_secure_installation was run."
+    exit
+  else
+    echo
+    echo "NO assigned ROOT password yet. RUNNING mysql_secure_installation..."
+    echo
+    run_mysql_secure_installation
+  fi
+else
+  # MySQL is NOT installed yet - so you have not yet changed the temporary
+  # password, so install MySQL. Then call run_mysql_secure_installation.
+  echo
+  echo "MySQL is NOT installed yet. GETTING and VERIFYING MySQL mysql80-community-release-el7-4."
+
+  # CHECK if MySQL package downloaded.
+  # Create a new file with the extension .rpm.md5 with contents of the copied
+  # MD5 from the approved site. With this file, perform an md5sum --check.
+  # If successful, returns exit code 0. If NOT successful, give user error
+  # message & returns exit code 1.
+  if ! [ -f mysql80-community-release-el7-4.noarch.rpm ]; then
+    wget https://dev.mysql.com/get/mysql80-community-release-el7-4.noarch.rpm
+  fi
+
+  if ! [ -f mysql80-community-release-el7-4.noarch.rpm.md5 ]; then
+    echo "8b55d5fc443660fab90f9dc328a4d9ad mysql80-community-release-el7-4.noarch.rpm" > mysql80-community-release-el7-4.noarch.rpm.md5
+  fi
+
+  # CHECK if retrieved MySQL package file passes md5sum check.
+  # CAN ALSO: Use GnuPG signatures to verify the integrity of downloaded packages.
+  if ! md5sum --check mysql80-community-release-el7-4.noarch.rpm.md5; then
+    # Not verifiable MySQL file. Show error here & exit from the script.
+    echo "FAILED md5sum check for the MySQL file. Exiting the script."
+    exit 1
+  fi
+  echo "GOT & VERIFIED MySQL package."
+
+  # CHECK if you can successfully query mysql80-community-release package, i.e.,
+  # 0 exit code - means package is already installed. BUT rpm command might FAIL.
+  # So for a clean install, erase the package & then install it.
+  echo
+  MYSQL_PKG="mysql80-community-release-el7-4.noarch.rpm"
+  if rpm -q "${MYSQL_PKG%.noarch.rpm}"; then
+    echo "Erasing previously installed MySQL package to ensure MySQL installation is clean."
+    sudo rpm -e "${MYSQL_PKG%.noarch.rpm}"
+  fi
+  echo
+  echo "NOW installing MySQL package & MySQL Server."
+  sudo rpm -ivh "${MYSQL_PKG}"
+  sudo yum install mysql-server -y
+  echo "DONE installing MySQL package with MySQL client & tools. DONE installing MySQL server."
+  echo
+fi
+
+# Must start mysqld.service to try running MySQL client.
+sudo systemctl start mysqld.service
+sudo systemctl status mysqld
+
+# CHECK CASE: At this point, MySQL is installed, mysql_secure_installation has
+# NOT been run, BUT mysqld.log file has lost temporary password.
+# Note: Without the if, when this line is run and fails, you are exited from script.
+# MYSQL_TEMP_PWD=$(sudo grep 'temporary password' /var/log/mysqld.log | awk '{print $13}')
+
+if ! MYSQL_TEMP_PWD=$(sudo grep 'temporary password' /var/log/mysqld.log | awk '{print $13}'); then
```

Comment:
Ah, you were saying you can never get this to fail -- I think you said you even tried deleting the `/var/log/mysqld.log` file.

One reason you might not be able to get it to fail could be that we need to put the command

```sh
set -o pipefail
```

at the top of the script where you run

```sh
set -e
```

This changes the shell so that in a command pipeline like

```sh
sudo grep ... | awk ...
```

if the first command exits with a non-zero status code, then the whole pipeline exits with that status code. By default, the pipeline will exit with the status code of the last command. I forgot to tell you about `set -o pipefail`, so the problems you are having here are probably my fault! Sorry!

Reply:
Combined line to: `set -eo pipefail` & re-tested.